Add Rego checks for Application & role Exclusions and be consistent across Conditional Access Policies

dagarwal-mitre opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #
  <!-- Remember this title will end up as the merge commit subject -->

## 🗣 Description ##
This PR modifies the Rego for all conditional access policies **prohibit role exclusions and application exclusions**. It also ensures that these **checks are applied in a consistent manner** across all 8 relevant policies. We removed the PolicyConditionsMatch ruleset because there was no easy way to consistently call it from all 8 policies due to minor differences in the checks so instead we opted for creating a common section of code that was placed into each policy.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
Closes #1170 
Closes #1323 

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
**These instructions are OBE because we are checking guest user exclusions anymore. Testing was performed with role exclusions and application exclusions**
To test, within the conditional access policies, inside of users, inside the ExternalGuestorExternalUsers, update the MembershipKind and the GuestOrExternalUserTypes, and update it to look like the image. The default for these is null, feel free to take a look at the AADBaseConfig.rego file to see the default around line 25. Test with the different tenants.

<img width="973" alt="Screenshot 2025-01-28 at 11 21 22 AM" src="https://github.com/user-attachments/assets/ca65fbd5-dfc2-4a61-8ee4-eee23e048dd3" />

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs may have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [x] PR passed smoke test check.
- [x] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [x] Resolved all merge conflicts on branch
- [x] Notified merge coordinator that PR is ready for merge via comment mention
- [ ] Demonstrate changes to the team for questions and comments. 
    (Note: Only required for issues of size `Medium` or larger)

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. This section is for the merge coordinator to complete. -->
- [x] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

twneale reviewed (commented):
Looks good 🐄

tkol2022 reviewed (changes requested):
Looks like some unit tests are missing for specific non-compliant states that PolicyConditionsMatch is checking for.

Missing unit tests for ExcludeRoles for 1.1, 3.1, 3.7, 3.8
Missing unit tests for ExcludeApplications for 1.1, 2.1, 2.3, 3.2, 3.6, 3.7, 3.8

tkol2022 commented:
The unit tests for policy 3.6 use an invalid JSON from AADBaseConfig.rego. In that JSON there is an IncludeUsers = All and then the unit tests add an IncludeRoles element as well. In a real tenant this can never happen because when you select the option to include roles you cannot then select the option to include all users as well. I think we should correct this to ensure that the policy and unit tests are aligned with practical scenarios. Given how finicky Rego is, we want to align the states we are testing with as closely to the real world as possible.

tkol2022 commented:
Some instances of calling PolicyConditionsMatch look like below whereas other cases the == true is omitted. Change them so they are all consistent to whatever you think is easier to maintain.

`PolicyConditionsMatch(CAPolicy, true) == true`

tkol2022 commented:
I think you should delete the line count(PrivRolesSet & ConvertToSet(CAPolicy.Conditions.Users.ExcludeRoles)) == 0 in AADConfig.rego for policy 3.6 because PolicyConditionsMatch verifies that the ExcludeRoles is empty so I think that line becomes OBE? I commented it out for now with my changes but take a look and see if you agree.

tkol2022 commented:
Make sure that all CAP policies have compliant and non-compliant unit tests which exercise User and Group exclusions via the config object. I found that policy 3.1 didn't have a compliant unit test for exclusions so I added it in my latest commit but check the other policy unit tests. This is especially important because we currently don't have a way to pass a config file to Invoke-ScubaCached so it's hard to even manually test this. Want to be sure we have coverage of these important scenarios.

tkol2022 commented:
> Some instances of calling PolicyConditionsMatch look like below whereas other cases the == true is omitted. Change them so they are all consistent to whatever you think is easier to maintain.
> 
> `PolicyConditionsMatch(CAPolicy, true) == true`

This is OBE because I removed PolicyConditionsMatch.

tkol2022 commented:
> count(PrivRolesSet & ConvertToSet(CAPolicy.Conditions.Users.ExcludeRoles))

This is no longer necessary because I removed it in my latest commit.

tkol2022 commented:
Take a look at policy MS.AAD.3.8v1
We use Count and count in the same ruleset. Can the line "count(FilterArray(Conditions, true)) > 0" be changed to upper case Count? I didn't want to modify it.

dagarwal-mitre commented:
I have addressed the comments, and have added new unit tests please re-review the unit tests and ensure they work as well as take a look at the AADrego file and check to make sure the CAPs are consistent.

tkol2022 commented:
> Looks like some unit tests are missing for specific non-compliant states that PolicyConditionsMatch is checking for.
> 
> Missing unit tests for ExcludeRoles for 1.1, 3.1, 3.7, 3.8 Missing unit tests for ExcludeApplications for 1.1, 2.1, 2.3, 3.2, 3.6, 3.7, 3.8

This was addressed.

tkol2022 commented:
> The unit tests for policy 3.6

I just pushed changes to address this.

tkol2022 commented:
> Make sure that all CAP policies have compliant and non-compliant unit tests which exercise User and Group exclusions via the config object. I found that policy 3.1 didn't have a compliant unit test for exclusions so I added it in my latest commit but check the other policy unit tests. This is especially important because we currently don't have a way to pass a config file to Invoke-ScubaCached so it's hard to even manually test this. Want to be sure we have coverage of these important scenarios.

@dagarwal-mitre I don't see unit tests for 3.6 in this regard. 
![image](https://github.com/user-attachments/assets/99fc67a8-6627-4cfc-97a7-33188d50cc6b)

tkol2022 commented:
> c

References noted to count were changed to be consistent.

tkol2022 commented:
> The unit tests for policy 3.6 use an invalid JSON from AADBaseConfig.rego. In that JSON there is an IncludeUsers = All and then the unit tests add an IncludeRoles element as well. In a real tenant this can never happen because when you select the option to include roles you cannot then select the option to include all users as well. I think we should correct this to ensure that the policy and unit tests are aligned with practical scenarios. Given how finicky Rego is, we want to align the states we are testing with as closely to the real world as possible.

I examined this again and determined that we will not make any changes at this time because the IncludeUsers field is not checked by the 3.6 policy, therefore it should have no bearing on the outcome of the policy evaluation. However I think it is bad practice to use JSON's that don't reflect what a real data structure would look like because it could have unintended side effects when evaluated in the rulesets. No action for this PR.

tkol2022 reviewed (changes requested):
The one pending changes is that we need unit tests for 3.6 that reference the config. See my latest comment on that in the PR thread.

twneale reviewed (approved):
The unit tests and y2zj1 run went fine for me. I'm not super conversant in what I'm looking at but overall it looks ok to me.

dagarwal-mitre commented:
@nanda-katikaneni this pr is now ready to merge, thanks!
